manifests/agent/windows.pp

```diff
@@ -1,73 +1,71 @@
-  $tempdir = 'C:/Users/distelli/AppData/Local/Temp/1',
-){
+# Install the Puppet Pipelines agent on Windows
+#
+# Do not use directly; use pipelines::agent.
+class pipelines::agent::windows {
+  $env = $facts['pipelines_env']
 
-  if $::facts['os']['hardware'] == 'x86_64' {
-    $archive = "distelli.Windows-AMD64-${version}.gz"
-    $url = "https://s3.amazonaws.com/download.distelli.com/distelli.Windows-AMD64/${archive}"
-  }
-  else {
-    $archive = "distelli.Windows-x86-${version}.gz"
-    $url = "https://s3.amazonaws.com/download.distelli.com/distelli.Windows-x86/${archive}"
-  }
-
-  $workdir = 'C:/Program Files/Distelli'
-
-  archive { "${tempdir}/${archive}" :
-    source       => $url,
-    # cleanup      => false,
-    creates      => "${tempdir}/distelli",
-    extract      => true,
-    extract_path => $tempdir,
-    require      => User['distelli'],
+  if ! $pipelines::agent::install_dir {
+    $install_dir = "${env['ProgramFiles']}\\Distelli"
+  } else {
+    $install_dir = $pipelines::agent::install_dir
   }
 
-  file { $workdir :
-    ensure  => directory,
-    owner   => 'distelli',
-    group   => 'Administrators',
-    require => User['distelli'],
+  if $pipelines::agent::version {
+    $download_url = "${pipelines::agent::download_url}/${pipelines::agent::version}.ps1"
+  } else {
+    $download_url = "${pipelines::agent::download_url}.ps1"
   }
+  $download_location = "${install_dir}\\distelli-download.ps1"
+  $download_cmd = "cmd.exe /c \"powershell -NoProfile -ExecutionPolicy Bypass -Command - < \"${download_location}\"\""
+  $agent_conf_file = "${env['SystemDrive']}\\distelli.yml"
 
-  # Requires fqdn_rand_string function from puppetlabs/stdlib
-  $distelli_exec = "distelli-${fqdn_rand_string(5)}.exe"
-
-  exec { 'Copy executable' :
-    command   => "Copy-Item ${tempdir}/distelli \$ENV:ProgramFiles/Distelli/${distelli_exec}",
-    unless    => "If (Test-Path -Path \$ENV:ProgramFiles/Distelli/${distelli_exec}) { exit 0 } else { exit 1}",
-    provider  => powershell,
-    require   => File[$workdir],
-    logoutput => true,
+  exec { "mkdir ${install_dir}":
+    command => "cmd.exe /c \"md \"${install_dir}\"\"",
+    path    => $facts['path'],
+    onlyif  => "cmd.exe /c \"if exist \"${install_dir}\" exit 1\"",
   }
-
-  file { ["${workdir}/distelli.exe",  "${workdir}/dagent.exe", "${workdir}/dtk.exe"] :
-    ensure  => link,
-    target  => "${workdir}/${distelli_exec}",
-    require => Exec['Copy executable'],
+  file { $download_location:
+    source  => $download_url,
+    require => Exec["mkdir ${install_dir}"],
   }
-
-  exec { 'Test distelli.exe execution' :
-    command     => "& \$ENV:ProgramFiles/Distelli/distelli.exe version",
+  exec { 'pipelines::agent download':
     provider    => powershell,
-    subscribe   => File["${workdir}/distelli.exe"],
-    logoutput   => true,
-    refreshonly => true,
+    require     => File[$download_location],
+    creates     => "${install_dir}\\distelli.exe",
```

Comment:
Hmm, I'm not sure that I would tie the download to checking for the binaries created post-install.

Think about how upgrades will be handled.

Reply:
The original intent here was to avoid running the install code if it is already installed. I thought that this "exec" would run if the dependency "require"d was updated (aka if the File[$download_location] changed... which would occur if a new version number was used). Is this a misunderstanding on my part? Is there a way to make it so this exec only runs if the File[$download_location] is changed?

Reply:
You should write a little manifest locally to verify for sure, but my understanding is exec won't ever run if the file pointed to by `creates` is present.

https://puppet.com/docs/puppet/5.3/types/exec.html#exec-attribute-creates

Reply:
Ah, I see what you mean! I've updated this so it uses:

```
    subscribe   => [
      File[$download_location],
    ],
    refreshonly => true,
```

...unfortunately, it seems that all commands are always re-executed on each puppet run though. This seems to be caused by the "mtime" of the $download_location being updated even though the content of that file is the same. Is there a way to make it so the mtime isn't updated here?

```
  file { $download_location:
    source  => $download_url,
    require => Exec["mkdir ${install_dir}"],
  }
```

Thanks!